miasm2/arch/mep/sem.py

```diff
@@ -0,0 +1,1147 @@
+    rn = rm ^ imm16
+
+
+### Shift instruction
+
+@sbuild.parse
+def sra(rn, rm_or_imm5):
+    """SRA - Shift Right signed"""
+
+    # Rn <- (Signed) Rn >> Rm4..0
+    # Rn <- (Signed) Rn >> imm5
+
+    # Unsigned result
+    shift_u = rn >> rm_or_imm5
+
+    # Signed result
+    shift_mask = i32(32) - rm_or_imm5
+    mask = (i32(0xFFFFFFFF) >> shift_mask) << shift_mask
+    shift_s = shift_u | mask
+
+    rn = shift_s if rn.msb() else shift_u
+
+
+@sbuild.parse
+def srl(rn, rm_or_imm5):
+    """SRL - Shift Right unsigned."""
+
+    # Rn <- (Unsigned) Rn >> Rm4..0
+    # Rn <- (Unsigned) Rn >> imm5
+    rn = rn >> rm_or_imm5
+
+
+@sbuild.parse
+def sll(rn, rm_or_imm5):
+    """SLL - Shift Left unsigned."""
+
+    # Rn <- (Unsigned) Rn >> Rm4..0
+    # Rn <- (Unsigned) Rn << imm5
+    rn = rn << rm_or_imm5
+
+
+@sbuild.parse
+def sll3(r0, rn, imm5):
+    """SLL3 - Shift Left unsigned, with 3 arguments."""
+
+    # R0 <- (Unsigned) Rn << imm5
+    r0 = rn << imm5
+
+
+@sbuild.parse
+def fsft(rn, rm):
+    "FSFT - Funnel shift."""
+
+    # Rn <- ((Rn||Rm)<<SAR5..0)63..32
+    # Note: lowest Rm bits are discarded
+
+    sar = SAR[:5].zeroExtend(32)
+    tmp_rn = rn << sar  # Shift Rn
+    tmp_rm = rm >> (i32(32) - sar)  # Shift Rm in the reverse order
+    rn = tmp_rn | tmp_rm  # Concatenate registers
+
+
+## Branch/Jump instructions
+
+@sbuild.parse
+def bra(disp12):
+    """BRA - Branch to an address."""
+
+    # PC <- PC + SignExt((disp12)11..1||0)
+    dst = disp12
+    PC = dst
+    ir.IRDst = dst
+
+
+@sbuild.parse
+def beqz(reg_test, disp8):
+    """BEQZ - Branch if the register stores zero."""
+
+    # if(Rn==0) PC <- PC +SignExt((disp8)7..1||0)
+    dst = ExprLoc(ir.get_next_break_loc_key(instr), 32) if reg_test else disp8
+    PC = dst
+    ir.IRDst = dst
+
+
+@sbuild.parse
+def bnez(reg_test, disp8):
+    """BNEZ - Branch if the register does not store zero."""
+
+    # if(Rn!=0) PC <- PC + SignExt((disp8)7..1||0)
+    dst = disp8 if "-"(reg_test, i32(0)) else ExprLoc(ir.get_next_break_loc_key(instr), 32)
+    PC = dst
+    ir.IRDst = dst
+
+
+@sbuild.parse
+def beqi(reg_test, imm4, disp16):
+    """BEQI - Branch if the register stores imm4."""
+
+    # if(Rn==ZeroExt(imm4)) PC <- PC +SignExt((disp17)16..1||0)
+    dst = ExprLoc(ir.get_next_break_loc_key(instr), 32) if "-"(reg_test, imm4) else disp16
+    PC = dst
+    ir.IRDst = dst
+
+
+@sbuild.parse
+def bnei(reg_test, imm4, disp16):
+    """BNEI - Branch if the register does not store imm4."""
+
+    # if(Rn!=ZeroExt(imm4)) PC <- PC+SignExt((disp17)16..1||0)
+    dst = disp16 if "-"(reg_test, imm4) else ExprLoc(ir.get_next_break_loc_key(instr), 32)
+    PC = dst
+    ir.IRDst = dst
+
+
+@sbuild.parse
+def blti(reg_test, imm4, disp16):
+    """BLTI - Branch if the register is lower than imm4."""
+
+    # if(Rn< ZeroExt(imm4)) PC <- PC +SignExt((disp17)16..1||0) - (Signed comparison)
+    dst = disp16 if "<"(reg_test, imm4) else ExprLoc(ir.get_next_break_loc_key(instr), 32)
```

Comment:
The failing tests are likely caused by this line. Locally, I am using PR #556.

What is the correct way to modify this line?

Reply:
Hi
First, Thanks for the PR !
For the sign problem: We plan to do a PR which introduce comparison operators, but it's not finish at that time. So if you can modify your PR to use current signed/unsigned equations, this is great! Don't worry, we will patch each architecture (including MeP once merged) to use the future API :smiley:

Is this ok for you?

Another point:
I started a review of the code: It seems you are introducing the use of `pytest`, which is currently not a dependance of Miasm. As we use regressions tests for commit checks, each machine will need `pytest`. Is it possible to modify the `test_all.py` in order to include MeP regressions tests to `test_all.py` to avoid the pytest introduction?

The point is that we made `test_all.py` quickly in order to solve the test regressions problem. If one day we plan to change it to use a real test module (it may be pytest!), we want a bit more time to test different "tests managers" before modifying the whole framework. Here, the PR 'forces' a bit the use of `pytest`. 
You may add a main calling your tests to fix this.

Reply:
The current version implements an abstraction layer in the `launch.py` that is compatible with `test_all.py` and `pytest`. The current drawback is that It duplicates the `launch_tests()` in 3 files =/
Does it sound ok ?

I did not manage to make the sign comparison work correctly.  With my local fixes (not pushed), the jit tests fail with `NotImplementedError: Unknown op: '<s'` in miasm2/ir/translators/C.py. I don't know how to fix it.

Reply:
Ok here is the trick:
First, you have to rip the flags computation from `expression/simplifications_cond.py`, for example in `expr_simp_inf_signed` for the signed inf.
Then, recode a function in your `sem.py` for example:
```python
def compute_s_inf(arg1, arg2):
    return ((arg1 - arg2) ^ ((arg1 ^ arg2) & ((arg1 - arg2) ^ arg1))).msb()
```

Then, add it to your sembuilder:
```python

def compute_s_inf(arg1, arg2):
    return ((arg1 - arg2) ^ ((arg1 ^ arg2) & ((arg1 - arg2) ^ arg1))).msb()

# SemBuilder context
ctx = {"PC": PC, "SP": SP, "LP": LP, "SAR": SAR, "TP": TP,
       "RPB": RPB, "RPE": RPE, "RPC": RPC, "EPC": EPC, "NPC": NPC,
       "EXC": EXC, "HI": HI, "LO": LO, "PSW": PSW, "DEPC": DEPC, "DBG": DBG,
       "exception_flags": exception_flags,
       "compute_s_inf":compute_s_inf}
sbuild = SemBuilder(ctx)
```

Once this, you can call the function in the sembuilder, and fix uage of 'high" comparison operators. For example:
```python

@sbuild.parse
def blti(reg_test, imm4, disp16):
    """BLTI - Branch if the register is lower than imm4."""

    # if(Rn< ZeroExt(imm4)) PC <- PC +SignExt((disp17)16..1||0) - (Signed comparison)
    dst = disp16 if compute_s_inf(reg_test, imm4) else ExprLoc(ir.get_next_break_loc_key(instr), 32)
    PC = dst
    ir.IRDst = dst

```


I know the equation is a bit large, but as I said you we will use the comparison operators in a next PR, but it's not ready yet :confused:

Is it ok for you?

Reply:
I think that I understood. What is the correct way to implement `compute_u_inf()` the unsigned counterpart?

Reply:
As I said, you can extract the equation from `expression/simplifications_cond.py`:
```python
def expr_simp_inf_unsigned_inversed(expr_simp, e):
    "((x - y) ^ ((x ^ y) & ((x - y) ^ x))) ^ x ^ y [31:32] == x <u y"
   ...
```

So you can write:
```python
def is_unsigned_inf(x, y):
    #  x <u y"
    result = (((x - y) ^ ((x ^ y) & ((x - y) ^ x))) ^ x ^ y).msb()
    return result
```

Reply:
Thanks! I did not understand that I needed to convert the equations myself. I pushed a second commit that adds the fixes. I could squash it later.

Two REPEAT jit examples fail and I do not understand why =/